src/jit/compiler.cpp

```diff
@@ -2478,6 +2479,28 @@ void Compiler::compSetProcessor()
 #endif // _TARGET_AMD64_
             }
         }
+#else  // !FEATURE_CORECLR
+        // If this is not FEATURE_CORECLR, the flags for SSE3, SSSE3, SSE41 and SSE42 are collectively
+        // implied by EnableSSE3_4, and, since we don't support the HW intrinsics, these and the AVX
+        // ISAs are the only ones supported by the VM and relied upon by the JIT.
+        //
+        if (JitConfig.EnableSSE3_4())
+        {
+            opts.setSupportedISA(InstructionSet_SSE3);
```

Comment:
I think, if we are going to meet the current check in `ee_il_dll.cpp` (https://github.com/dotnet/coreclr/blob/master/src/jit/ee_il_dll.cpp#L408), we need to also have a separate check there (likely one that just checks `if ((JitConfig.EnableAVX2() && JitConfig.EnableAVX() && (JitConfig.EnableSSE3_4() != 0))`.

We might also want to, for correctness sake (although I don't think it will be impactful to Desktop) ensure that `ops.setSupportedISA` is called for `InstructionSet_SSE` and `InstructionSet_SSE2`...

Reply:
I think the check in ee_il_dll.cpp should be checking for **either** HW intrinsics being enabled **or** `featureSIMD`. I don't think disabling HW intrinsics should prevent 32-byte `Vector<T>`.

Reply:
In fact, it's unclear to me why that check in ee_il_dll.cpp needs the final condition (`JitConfig.EnableHWIntrinsic() != 0`) at all.

Reply:
> In fact, it's unclear to me why that check in ee_il_dll.cpp needs the final condition (JitConfig.EnableHWIntrinsic() != 0) at all.

The final check is there because:
1. `canUseVexEncoding()` is tied to the compiler support for AVX
2. The HWIntrinsics are hierarchical (disabling the base HWIntrinsics also disables the SSE-AVX2 HWIntrinsics)

Ideally, `canUseVexEncoding()` would instead be tied to:
1. The VM indicating the CPU supports AVX, BMI1, or BMI2 (AVX2, FMA, also use VEX but are covered by the AVX check)
2, The EnableVEX flag (which controls the compiler support for emitting the VEX prefix).

Then, the `ee_il_dll` code for the `Vector<T>` length could be a check for:
1. The VM indicating the CPU supports AVX2
2. The `VectorTSize` flag either being 0 (default) or a valid power of 2 greater than `32`

(On desktop, the code likely needs to stay as `(JitConfig.EnableAVX2() != 0) && (JitConfig.EnableAVX() != 0)`, as it was before, since we won't be flowing the new VM flags back)

Reply:
This **must** change. It's unacceptable that the `Vector<T>` support would be disabled by disabling HW intrinsic support - especially since that's a relatively new feature, and I think we must have a way to disable it in case, post-release, some issue is discovered that doesn't impact SNV.dll.

Aside from what you say above, I don't think that disabling HW intrinsics should in any way impact what ISAs are considered supported.

Reply:
The quickest fix, to break that dependency, is to probably change `ee_il_dll.cpp` to the following (which drops the `EnableHWIntrinsic` check for CoreCLR and makes desktop match what it was doing before):

```cpp
#ifdef FEATURE_CORECLR
    if ((JitConfig.EnableAVX2() != 0) && (JitConfig.EnableAVX() != 0) && (JitConfig.EnableSSE42() != 0) &&
        (JitConfig.EnableSSE41() != 0) && (JitConfig.EnableSSSE3() != 0) && (JitConfig.EnableSSE3() != 0) &&
        (JitConfig.EnableSSE2() != 0) && (JitConfig.EnableSSE() != 0))
#else
    if ((JitConfig.EnableAVX2() != 0) && (JitConfig.EnableAVX() != 0))
#endif
```

We can then change the `FEATURE_CORECLR` codepath in `compiler.cpp` to do:

```cpp
if (JitConfig.EnableHWIntrinsic())
{
    opts.setSupportedISA(InstructionSet_Base);
}

if (JitConfig.EnableSSE())
{
    // ..., rest of the checks, in hierarchical order
}
```

There might be a little bit of additional cleanup for @fiigii or I to do in `hwintrinsicxarch.cpp`, but we don't have any CI legs testing the `EnableHWIntrinsic` flag right now, so I think we can do it in a followup

Reply:
> change ee_il_dll.cpp to the following (which drops the EnableHWIntrinsic check for CoreCLR and makes desktop match what it was doing before):

Yes, that's the changed I've made (which addresses the desktop issue where it was not only limiting the size of `Vector<T>` when `COMPlus_EnableHWIntrinsic == 0` but was also asserting.

I'll make the additional change you suggest.